src/app/system/manage-data-tables/edit-data-table/edit-data-table.component.ts

```diff
@@ -0,0 +1,297 @@
+   * @returns {string} Column Type.
+   */
+  getColumnType(columnDisplayType: string): string {
+    if (columnDisplayType === '') {
+      return '';
+    } else if (columnDisplayType === 'INTEGER') {
+      return 'Number';
+    } else if (columnDisplayType === 'CODELOOKUP') {
+      return 'Dropdown';
+    } else {
+      return columnDisplayType[0] + columnDisplayType.substr(1).toLowerCase();
+    }
+  }
+
+  /**
+   * Gets the columns form array.
+   * @returns {FormArray} Columns form array.
+   */
+  get columns(): FormArray {
+    return this.dataTableForm.get('columns') as FormArray;
+  }
+
+  /**
+   * Adds the column form with the given values to given column form array.
+   * Calls onNewColumnChanges to subscribe for the value changes in all of the fields of that column row.
+   * @param {FormArray} columnFormArray Given column form array.
+   * @param {string} columnName Column Name.
+   * @param {string} columnType Column Type.
+   * @param {string} columnLength Column Length.
+   * @param {string} columnCode Column Code.
+   */
+  addColumn(columnFormArray: FormArray) {
+    columnFormArray.push(this.createColumnForm('', '', '', '', false));
+    this.onNewColumnChanges(columnFormArray, columnFormArray.length - 1);
+  }
+
+  /**
+   * Removes the column form from given column form array at given index.
+   * Initalizes dropColumns array, if it is not already.
+   * Pushes name of the column to dropColumns array.
+   * @param {FormArray} columnFormArray Given column form array.
+   * @param {number} index Array index from where column form needs to be removed.
+   */
+  removeColumn(columnFormArray: FormArray, index: number) {
+    if (this.dataTableChangesData.dropColumns === undefined && columnFormArray.at(index).get('name').value !== '') {
+      this.dataTableChangesData.dropColumns = [];
+    }
+    if (columnFormArray.at(index).get('name').value !== '') {
+      this.dataTableChangesData.dropColumns.push({ name: columnFormArray.at(index).get('name').value });
+    }
+    columnFormArray.removeAt(index);
+  }
+
+  /**
+   * Enables Length Text Field and disables Code Text Field when Column Type is 'String'.
+   * Enables Code Text Field and disables Length Text Field when Column Type is 'Dropdown'.
+   * Disables both Length and Code Text Field when Column Type is neither 'String' nor 'Dropdown'.
+   * Initalizes addColumns array, if it is not already.
+   * Subscribes to the name, type, mandatory, length and code field of respective column.
+   * @param {number} index Index for Column Form Array.
+   * @param {FormArray} columnFormArray Column Form Array.
+   */
+  onNewColumnChanges(columnFormArray: FormArray, index: number) {
+    if (this.dataTableChangesData.addColumns === undefined) {
+      this.dataTableChangesData.addColumns = [];
+    }
+    this.dataTableChangesData.addColumns.push({});
+    const i = this.dataTableChangesData.addColumns.length - 1;
+    columnFormArray.at(index).get('name').valueChanges
```

Comment:
Isn't it better to subscribe to value changes of the form itself and replace the entire column with the new value instead? Will that cause any issue?

Reply:
We cannot subscribe to the entire form, it will then come with loads of if-statements which will make it harder to test.

Reply:
In case of add, we can replace the entire column as I mentioned so no if statements would be required.

Reply:
If we subscribe to the entire column, then we to put a switch statement in order to identify which field's value is changed.

Reply:
It is not needed, we can just replace the entire column as this is adding a new one.